docs/tutorial/from-zero-to-hero-write-your-first-kubernetes-charm/integrate-your-charm-with-postgresql.md

```diff
@@ -359,6 +359,166 @@ This should produce something similar to the output below (of course, with the n
+
+    assert state_out.unit_status == ops.BlockedStatus("Waiting for database relation")
+```
+
+Run `tox -e unit` to make sure all test cases pass.
+
+## Add integration tests for your charm
+
+Now that our charm integrates with the PostgreSQL database, without a database integration, the app will be in the blocked status instead of active, because it requires a database relation. Let's tweak our existing integration test case `test_build_and_deploy` accordingly:
+
+```python
+async def test_build_and_deploy(ops_test: OpsTest):
+    """Build the charm-under-test and deploy it together with related charms.
+
+    Assert on the unit status before any relations/configurations take place.
+    """
+    # Build and deploy charm from local source folder
+    charm = await ops_test.build_charm(".")
+    resources = {
+        "demo-server-image": METADATA["resources"]["demo-server-image"]["upstream-source"]
+    }
+
+    # Deploy the charm and wait for blocked/idle status
+    # The app will not be in active status as this requires a database relation
+    await asyncio.gather(
+        ops_test.model.deploy(charm, resources=resources, application_name=APP_NAME),
+        ops_test.model.wait_for_idle(
+            apps=[APP_NAME], status="blocked", raise_on_blocked=False, timeout=300
+        ),
+    )
+```
+
+Then, let's add another test case to check the integration is successful. For that, we need to deploy a database to the test cluster and integrate both applications. If everything works as intendedFinally, we should check that the charm reports an active status.
+
+In your `tests/integration/test_charm.py` file add the following test case:
+
+```python
+@pytest.mark.abort_on_fail
+async def test_database_integration(ops_test: OpsTest):
+    """Verify that the charm integrates with the database.
+
+    Assert that the charm is active if the integration is established.
+    """
+    await ops_test.model.deploy(
+        application_name="postgresql-k8s",
+        entity_url="postgresql-k8s",
+        channel="14/stable",
+    )
+    await ops_test.model.integrate(f"{APP_NAME}", "postgresql-k8s")
+    await ops_test.model.wait_for_idle(
+        apps=[APP_NAME], status="active", raise_on_blocked=False, timeout=300
+    )
+
+```
+
+```{important}
+
+But if you run the one and then the other (as separate `pytest ...` invocations, then two separate models will be created unless you pass `--model=some-existing-model` to inform pytest-operator to use a model you provide.
+
+```
+
+In your Multipass Ubuntu VM, run the test again:
+
+```text
+ubuntu@charm-dev:~/fastapi-demo$ tox -e integration
+```
+
+The test may again take some time to run.
+
+```{tip}
+
+**Pro tip:** To make things faster, use the `--model=<existing model name>` to inform `pytest-operator` to use the model it has created for the first test. Otherwise, charmers often have a way to cache their pack or deploy results; an example is [spellbook](https://github.com/canonical/spellbook).
```

Comment:
Dima made a suggestion in another PR to get rid of spellbook here - I think that's reasonable. Also, I don't think we need the "Pro tip" bit when this is already in a tip box.

```suggestion
To make things faster, use the `--model=<existing model name>` to inform `pytest-operator` to use the model it has created for the first test. Otherwise, charmers often have a way to cache their pack or deploy results.
```